patches/net/minecraft/world/level/saveddata/SavedData.java.patch

```diff
@@ -1,15 +1,20 @@
 --- a/net/minecraft/world/level/saveddata/SavedData.java
 +++ b/net/minecraft/world/level/saveddata/SavedData.java
-@@ -37,7 +_,7 @@
+@@ -36,18 +_,23 @@
+             compoundtag.put("data", this.save(new CompoundTag(), p_324088_));
              NbtUtils.addCurrentDataVersion(compoundtag);
  
++            net.neoforged.neoforge.common.IOUtilities.withIOWorker(()->{
```

Comment:
What if the data is already being saved when this is called? Should we care?

Reply:
For what it's worth, in 1.21.4 vanilla keeps a `CompletableFuture` around to which a new future is appended with `thenCompose()` when a new set of `SavedData` is written to disk. The future being appended encompasses either a future per data-to-be-saved or a future per some group of data-to-be-saved depending on whether the amount of `SavedData`s to be written exceeds `Util.maxAllowedExecutorThreads()` or not. This way the saving is parallelized per set of `SavedData` to be written at once but multiple subsequent attempts at writing a certain `SavedData` are guaranteed to be serial.

Reply:
> What if the data is already being saved when this is called? Should we care?

At worst it will just be queued up to write again. This is another reason I used a single threaded executor, in addition to the fact that it's all going to be on the file system so concurrent writing would provide little benefit.